src/python/pants/backend/python/rules/python_test_debugger.py

```diff
@@ -0,0 +1,116 @@
+from pants.rules.core.strip_source_root import SourceRootStrippedSources
+from pants.util.contextutil import temporary_dir
+
+
+@rule(name="Run pytest in an interactive process")
+async def debug_python_test(
+  test_target: PythonTestsAdaptor,
+  pytest: PyTest,
+  python_setup: PythonSetup,
+  build_root: BuildRoot,
+  runner: InteractiveRunner,
+  workspace: Workspace,
+) -> TestDebugResult:
+
+  transitive_hydrated_targets = await Get[TransitiveHydratedTargets](
+    BuildFileAddresses((test_target.address,))
+  )
+  all_targets = transitive_hydrated_targets.closure
+  all_target_adaptors = [t.adaptor for t in all_targets]
+
+  interpreter_constraints = PexInterpreterConstraints.create_from_adaptors(
+    adaptors=tuple(all_target_adaptors),
+    python_setup=python_setup
+  )
+
+  requirements = PexRequirements.create_from_adaptors(
+    adaptors=all_target_adaptors,
+    additional_requirements=pytest.get_requirement_strings()
+  )
+
+  source_root_stripped_test_target_sources = await Get[SourceRootStrippedSources](
+    Address, test_target.address.to_address()
+  )
+
+  source_root_stripped_sources = await MultiGet(
+    Get[SourceRootStrippedSources](HydratedTarget, hydrated_target)
+    for hydrated_target in all_targets
+  )
+
+  stripped_sources_digests = tuple(
+    stripped_sources.snapshot.directory_digest for stripped_sources in source_root_stripped_sources
+  )
+  sources_digest = await Get[Digest](DirectoriesToMerge(directories=stripped_sources_digests))
+
+  output_pytest_requirements_pex = 'pytest-with-requirements.pex'
+
+  resolved_requirements_pex = await Get[Pex](
+    CreatePex(
+      output_filename=f'./{output_pytest_requirements_pex}',
+      requirements=requirements,
+      interpreter_constraints=interpreter_constraints,
+      entry_point="pytest:main",
+    )
+  )
+
+  inits_digest = await Get[InjectedInitDigest](Digest, sources_digest)
+
+  merged_input_files = await Get(
+    Digest,
+    DirectoriesToMerge(
+      directories=(
+        sources_digest,
+        inits_digest.directory_digest,
+        resolved_requirements_pex.directory_digest,
+      )
+    ),
+  )
+
+  with temporary_dir(root_dir=str(Path(build_root.path, ".pants.d")), cleanup=True) as tmpdir:
+    path_relative_to_build_root = Path(tmpdir).relative_to(build_root.path)
+    test_target_sources_file_names = sorted(
+      f'{path_relative_to_build_root}/{snapshot_file}'
+      for snapshot_file
+      in source_root_stripped_test_target_sources.snapshot.files
+    )
+    pex_args = (*pytest.get_args(), *test_target_sources_file_names)
+
+    workspace.materialize_directory(
+      DirectoryToMaterialize(merged_input_files, path_prefix=path_relative_to_build_root)
+    )
+
+    request_args = [f'{path_relative_to_build_root}/{output_pytest_requirements_pex}', *pex_args]
+    run_request = InteractiveProcessRequest(argv=request_args, run_in_workspace=True)
+
+    result = runner.run_local_interactive_process(run_request)
```

Comment:
This pattern mirrors how we use `tmpdir` and the interactive runner in the `run` rule, unsure if we need to wrap this in a `try` / `except` (as done there) or if `.pants.d` is the right spot for the temporary directory - calling that out in case folks have further insights.

Reply:
I don't think the tmpdir necessarily needs to be under .pants.d, I think it can be a regular tmpdir. Anyone else have thoughts?

Reply:
The reason it goes under `.pants.d` is that `workspace_materialize_directory` is only capable of writing relative to the build root. If you need to materialize to disk, then you'd use `.pants.d` because it's gitignored (in case cleanup fails for whatever reason). 

@gshuflin is it actually necessary to materialize the PEX to the build root for IPR to work?

Reply:
@Eric-Arellano no, the PEX needs to be materialized to whatever the InteractiveProcessRequest's working directory is, which is either a tmpdir or the build root. However, we don't currently have an abstraction that will let us materialize files to the tmpdir that IPR creates if `run_in_workspace` is False. https://github.com/pantsbuild/pants/pull/8855 is an attempt to provide that abstraction.